components/web_discovery/browser/payload_generator.cc

```diff
@@ -0,0 +1,203 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/web_discovery/browser/payload_generator.h"
+
+#include <utility>
+
+#include "base/containers/fixed_flat_set.h"
+#include "brave/components/web_discovery/browser/privacy_guard.h"
+
+namespace web_discovery {
+
+namespace {
+
+constexpr char kCountryCodeFieldName[] = "ctry";
+constexpr char kSearchResultKey[] = "r";
+constexpr char kSearchResultURLKey[] = "u";
+constexpr size_t kMinSearchResultSize = 4;
+
+constexpr char kAliveAction[] = "alive";
+constexpr char kStatusFieldName[] = "status";
+constexpr char kTimestampFieldName[] = "t";
+
+constexpr auto kQueryActions = base::MakeFixedFlatSet<std::string_view>(
+    {"query", "anon-query", "widgetTitle"});
+
+bool ValueHasContent(const base::Value& value) {
+  const auto* value_str = value.GetIfString();
+  if (value_str) {
+    return !value_str->empty();
+  }
+  if (!value.is_none()) {
+    return true;
+  }
+  return false;
+}
+
+bool AggregatedDictHasContent(const base::Value::Dict& dict) {
+  for (const auto [_k1, value] : dict) {
+    const auto* value_dict = value.GetIfDict();
+    if (!value_dict) {
+      continue;
+    }
+    for (const auto [_k2, sub_value] : *value_dict) {
+      if (ValueHasContent(sub_value)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool IsPrivateResult(const PayloadRule& rule,
+                     const PatternsURLDetails* matching_url_details,
+                     const base::Value::Dict& dict) {
+  if (rule.key != kSearchResultKey) {
+    return false;
+  }
+  const auto* url = dict.FindString(kSearchResultURLKey);
+  if (!url) {
+    return false;
+  }
+  return IsPrivateURLLikely(GURL(*url), matching_url_details);
+}
+
+bool ShouldDropSearchResultPayload(const PayloadRule& rule,
+                                   size_t result_size) {
+  if (rule.key != kSearchResultKey) {
+    return false;
+  }
+  return result_size < kMinSearchResultSize;
+}
+
+base::Value::Dict CreatePayloadDict(const PayloadRuleGroup& rule_group,
+                                    base::Value::Dict inner_payload) {
+  base::Value::Dict payload;
+  payload.Set(kActionKey, rule_group.action);
+  payload.Set(kInnerPayloadKey, std::move(inner_payload));
+  return payload;
+}
+
+std::optional<base::Value> GenerateClusteredJoinedPayload(
+    bool is_query_action,
+    const PayloadRule& rule,
+    const PatternsURLDetails* matching_url_details,
+    const std::vector<base::Value::Dict>& attribute_values) {
```

Comment:
how large this one can be? this processing is probably a candidate for moving to the task runner